project/utils/include/utils/cuda.cuh

```diff
@@ -0,0 +1,247 @@
+     * 
+     * @param[in] kernel         The kernel to be launched.
+     * @param[in] threads_needed The number of threads that need to be launched.
+     * @param[in] launch_kernel  A function to launch the kernel.
+     * 
+     * @c threads_needed does not need to be rounded; it can just be the size of
+     * the data. This function will perform any appropriate rounding/padding.
+     * However, note that either @c launch_kernel or @c kernel will need to
+     * perform checks to make sure that @c kernel does not perform an 
+     * out-of-bounds memory access in case more threads are launched than are
+     * needed.
+     * 
+     * @throws CudaException if a CUDA error occurs.
+     */
+    template<typename KernelFunction>
+    void launch_kernel(KernelFunction kernel, 
+                       size_t threads_needed, 
+                       const KernelLaunchFunction& launch_kernel)
+    {
+        // Determine the ideal grid and block size.
+        int min_grid_size;
+        int block_size;
+        CUDA_CALL(cudaOccupancyMaxPotentialBlockSize(&min_grid_size, 
+                                                     &block_size, 
+                                                     kernel, 
+                                                     0, 
+                                                     0));
+        
+        // Calculate the actual grid size needed to launch each of the threads.
+        auto grid_size = (threads_needed + block_size - 1) / block_size;
+
+        // Call the kernel.
+        launch_kernel(0, 
+                      static_cast<unsigned>(grid_size), 
+                      static_cast<unsigned>(block_size));
+    }
+
+    //--------------------------------------------------------------------------
+    // Memory Management
+    //--------------------------------------------------------------------------
+
+    /// Type of a function pointer for a CUDA malloc function.
+    using CudaMalloc = cudaError_t (*)(void** ptr, size_t size);
+
+    /// Type of a function pointer for a CUDA free function.
+    using CudaFree = cudaError_t (*)(void* ptr);
+
+    /**
+     * Memory managed wrapper for a memory region allocated via CUDA.
+     * 
+     * @tparam T      The type of element held in the memory region.
+     * @tparam MALLOC The CUDA function used to allocate the memory region.
+     * @tparam FREE   The CUDA function used to clear the memory region.
+     */
+    template<typename T, CudaMalloc MALLOC, CudaFree FREE>
+    class CudaBuffer
+    {
+    public:
+        /**
+         * Construct a CudaBuffer with the specified number of elements.
+         * 
+         * @param num_elements The number of elements to allocate.
+         *
+         * @throw std::bad_alloc If the memory allocation fails.
+         */
+        explicit CudaBuffer(size_t num_elements) :
+            num_elements(num_elements),
+            pointer(malloc(size()), FREE)
```

Comment:
Honestly, I didn't really think of this, probably because I don't know that I've ever really seen cudaFree() fail. Looking at the API docs, it only seems to fail in the case of a double free or if passed a pointer that was never allocated properly.

But like you said, even if I checked the return value, I can't really do much about it, since the deleter is called from std::unique_ptr's destructor. I could, at best, log the error.